### PR TITLE
Set buffer size in bytes

### PR DESCRIPTION
**Pull request by grosser**
Default buffer size matches dd-agent read buffer

@Antti @jonmoter 

as replacement for https://github.com/DataDog/dogstatsd-ruby/pull/82 and https://github.com/DataDog/dogstatsd-ruby/pull/65

### CODE REVIEW

**Comment by grosser** [comment]
@gmmeyer @masci @degemer

File: lib/datadog/statsd.rb [hunk 1]
```diff
@@ -93,16 +93,19 @@ def self.VERSION
     # @param [Integer] port your statsd port
     # @option opts [String] :namespace set a namespace to be prepended to every metric name
     # @option opts [Array<String>] :tags tags to be added to every metric
-    # @option opts [Integer] :max_buffer_size max messages to buffer
+    # @option opts [Integer] :max_buffer_bytes max messages to buffer
```
Comment (grosser) [inline lib/datadog/statsd.rb]: this option was never released, so it should be fine to change it
users that used the additional argument before will get a "ArgumentError" and then have to read the docs on how to upgrade

File: lib/datadog/statsd.rb [hunk 2]
```diff
@@ -443,17 +446,28 @@ def send_stats(stat, delta, type, opts=EMPTY_OPTIONS)
...
       end
     end
 
     def flush_buffer
-      return @buffer if @buffer.empty?
-      send_to_socket(@buffer.join(NEW_LINE))
-      @buffer = Array.new
+      return if @buffer_bytes.zero?
+      send_to_socket(@buffer)
+      @buffer = String.new
```
Comment (grosser) [inline lib/datadog/statsd.rb]: not using `""` since that breaks the tests on ruby 2.3+ where we use frozen strings

**Comment by gmmeyer** [comment]
Hey @grosser I'll merge this after you rebase, this looks pretty good to me. Do you mind if I close #82 since this replaces it?

**Comment by grosser** [comment]
rebased!
can close 82 and 64

**Comment by grosser** [comment]
after that the only missing things is thread-safety, that can be another minor release or just added to 4.0

**Comment by grosser** [comment]
thread safe batching will need some rework to not be a big mess ... so prefer getting 4.0 out ...
... afaik multi-threaded udp sending already works fine ... we used that for a few years without issues

File: lib/datadog/statsd.rb [hunk 3]
```diff
@@ -432,17 +434,28 @@ def send_stats(stat, delta, type, opts=EMPTY_OPTIONS)
 
     def send_stat(message)
       if @batch_nesting_depth > 0
+        message_bytes = message.bytesize
```
Comment (humzashah) [inline lib/datadog/statsd.rb]: I think this leaves open the (unlikely) possibility of `message_bytes` being greater than `@max_buffer_size` and us ending up with the same problem. What do you think about raising an error if that ever happens? 

```ruby
if message_bytes >= @max_buffer_bytes
  raise ArgumentError.new("Message size is too large.")
end
```

File: lib/datadog/statsd.rb [hunk 4]
```diff
@@ -108,8 +108,10 @@ def initialize(host = DEFAULT_HOST, port = DEFAULT_PORT, opts = EMPTY_OPTIONS)
       raise ArgumentError, 'tags must be a Array<String>' unless tags.nil? or tags.is_a? Array
       @tags = (tags || []).compact.map! {|tag| escape_tag_content(tag)}
 
-      @buffer = Array.new
-      @max_buffer_size = opts[:max_buffer_size] || 50
+      # batching
+      @max_buffer_bytes = opts[:max_buffer_bytes] || 8192
```
Comment (humzashah) [inline lib/datadog/statsd.rb]: What are your thoughts on warning somebody when they set `max_buffer_bytes` greater than this default?

File: lib/datadog/statsd.rb [hunk 5]
```diff
@@ -432,17 +434,28 @@ def send_stats(stat, delta, type, opts=EMPTY_OPTIONS)
 
     def send_stat(message)
       if @batch_nesting_depth > 0
+        message_bytes = message.bytesize
+        unless @buffer_bytes.zero?
```
Comment (humzashah) [inline lib/datadog/statsd.rb]: IMO `if @buffer_bytes != 0` would be much clearer. Admittedly, this is subjective.

**Comment by grosser** [comment]
I think the agent can be configured to accept higher values, so that should
be fine.
A warning for something that might be a valid usecase also means we must
allow users to opt-out of the warning,
if someone sets this option I'd trust them to know what they are doing.

File: lib/datadog/statsd.rb (hunk 5 above)
Comment (grosser) [reply lib/datadog/statsd.rb]: yeah, updated

**Comment by grosser** [comment]
needs https://github.com/DataDog/dogstatsd-ruby/pull/93 to be merged first

**Comment by grosser** [comment]
@gmmeyer rebased and green 🎉

**Comment by gmmeyer** [comment]
I agree with your assessment @grosser that the raise wouldn't be necessary for this, that is configurable on the agent size

**Comment by grosser** [comment]
maybe good enough for a 4.0 release ...
thread-safety can wait :D

**Pull request merged**